### PR TITLE
✨ Add `caphv-use=allow` tag for device usage

### LINKED ISSUE #115: Add tag to mark devices that can be chosen by CAPHV
/kind proposal

***Goals***
Mark devices in an easy and concise way that can be chosen by CAPHV to be added to a cluster

**User Story**
As a user, I want to easily tag those devices, that I want to include in my CAPI clusters. The other devices that are not tagged explicitly, should not be chosen. In this way, I am sure to not accidentally provision devices with CAPI that I use for other purposes. 

**Detailed Description**
Add the tag "caphv-use=allow" to our hvtag library and include a function hvtag.DeviceUsableByCAPI() bool to search for this tag and to return whether it exists (in this form) or not. Ignore all other tags that are similar but maybe not the same.

**Contract changes**
This is related to the HivelocityDeviceType which should be replaced according to #114. Ignore all changes to HivelocityDeviceType in this issue to not mix the two topics.

### PR DESCRIPTION
**Pull request by aniruddha2000**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #115 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: pkg/services/hivelocity/hvtag/hvtag.go [hunk 1]
```diff
@@ -42,6 +42,9 @@ const (
 
 	// DeviceTagKeyPermanentError is the key for machines which need a manual reset by a Hivelocity admin.
 	DeviceTagKeyPermanentError DeviceTagKey = "caphv-permanent-error"
+
+	// DeviceTagKeyClusterUse is the key to allow device use by CAPI cluster.
+	DeviceTagKeyClusterUse DeviceTagKey = "caphv-use"
```
Comment (janiskemper) [inline pkg/services/hivelocity/hvtag/hvtag.go:47]: name it DeviceTagKeyCAPHVUseAllowed

File: pkg/services/hivelocity/hvtag/hvtag.go [hunk 2]
```diff
@@ -90,7 +94,7 @@ func DeviceTagFromList(key DeviceTagKey, tagList []string) (DeviceTag, error) {
 			continue
 		}
 
-		// get DeviceTag from tagString
+		// get DeviceTag from tagStringdeviceTagFromString
```
Comment (janiskemper) [inline pkg/services/hivelocity/hvtag/hvtag.go:97]: ??

File: pkg/services/hivelocity/hvtag/hvtag.go [hunk 3]
```diff
@@ -135,6 +139,16 @@ func PermanentErrorTagFromList(tagList []string) (DeviceTag, error) {
 	return DeviceTagFromList(DeviceTagKeyPermanentError, tagList)
 }
 
+// DeviceUsableByCAPI returns if cluster can use the device.
+func DeviceUsableByCAPI(tagList []string) (bool, error) {
```
Comment (janiskemper) [inline pkg/services/hivelocity/hvtag/hvtag.go:143]: this function should return a boolean and no error. The question is just yes or no. If something is wrong, then it's a no.

**Review by janiskemper — CHANGES_REQUESTED** [review]

**Review by janiskemper — APPROVED** [review]

**Review by janiskemper — APPROVED** [review]

**Comment by janiskemper** [comment]
@guettli plz have a look at the docs and approve if they match your expectations

File: docs/book/src/developer/setup.md [hunk 4]
```diff
@@ -2,12 +2,12 @@
 
 ## Tagging Machines
 
-First you need to tag machines, so that the CAPHV controller can select and provision them.
+First you need to tag machines, so that the CAPHV controller can select and provision them. The devices must have `caphv-use=allow` tag so that the controller can use them.
 
 See [Provisioning Machines](../topics/provisioning-machines.md) for more about this.
```
Comment (guettli) [inline docs/book/src/developer/setup.md:7]: @aniruddha2000 please add this to the file provisioning-machines.md, too.

**Review by guettli — APPROVED** [review]
lgtm.

**Comment by aniruddha2000** [comment]
@guettli the e2e fails, is there anything else need changes

**Comment by janiskemper** [comment]
I looked into the e2e test dump to understand myself what might be going on, but it looks like the server was provisioned. @guettli is this server maybe stuck now in some way?

```
{"level":"DEBUG","time":"2023-11-20T08:14:48.610Z","file":"device/device.go:645","message":"Completed function","controller":"hivelocitymachine","controllerGroup":"infrastructure.cluster.x-k8s.io","controllerKind":"HivelocityMachine","HivelocityMachine":{"name":"caphv-3r9grv-md-0-5nhvp","namespace":"caphv-54cb2q"},"namespace":"caphv-54cb2q","name":"caphv-3r9grv-md-0-5nhvp","reconcileID":"b4104766-4a0d-4e10-9073-5c2a5405fa75","HivelocityMachine":{"name":"caphv-3r9grv-md-0-5nhvp","namespace":"caphv-54cb2q"},"Machine":{"name":"caphv-3r9grv-md-0-595ccf6bf7x6656b-67rnr","namespace":"caphv-54cb2q"},"Cluster":{"name":"caphv-3r9grv","namespace":"caphv-54cb2q"},"HivelocityCluster":{"name":"caphv-3r9grv","namespace":"caphv-54cb2q"},"function":"actionDeviceProvisioned","DeviceId":14730,"PowerStatus":"ON","script":"#cloud-config\n## template: jinja\n#cloud-config\n\nwr"}
{"level":"DEBUG","time":"2023-11-20T08:14:48.611Z","file":"device/state_machine.go:84","message":"ReconcileState","controller":"hivelocitymachine","controllerGroup":"infrastructure.cluster.x-k8s.io","controllerKind":"HivelocityMachine","HivelocityMachine":{"name":"caphv-3r9grv-md-0-5nhvp","namespace":"caphv-54cb2q"},"namespace":"caphv-54cb2q","name":"caphv-3r9grv-md-0-5nhvp","reconcileID":"b5164c3a-8149-41f4-89b8-c8f40b208dff","HivelocityMachine":{"name":"caphv-3r9grv-md-0-5nhvp","namespace":"caphv-54cb2q"},"Machine":{"name":"caphv-3r9grv-md-0-595ccf6bf7x6656b-67rnr","namespace":"caphv-54cb2q"},"Cluster":{"name":"caphv-3r9grv","namespace":"caphv-54cb2q"},"HivelocityCluster":{"name":"caphv-3r9grv","namespace":"caphv-54cb2q"},"caller":"state-machine","initialState.ProvisioningState":"provisioned"}
```

**Comment by guettli** [comment]
I look into the logs.

If I remove debug-level logs, then these changes are the last before the cluster got deleted:


```
level: INFO
time: "2023-11-20T07:57:08.555Z"
file: client/hvclient.go:167
message: calling ProvisionDevice()
controller: hivelocitymachine
controllerGroup: infrastructure.cluster.x-k8s.io
controllerKind: HivelocityMachine
HivelocityMachine:
  name: caphv-3r9grv-md-0-5nhvp
  namespace: caphv-54cb2q
namespace: caphv-54cb2q
name: caphv-3r9grv-md-0-5nhvp
reconcileID: 662fd9c2-e3f5-4af3-9796-6fd29b5999b7
HivelocityMachine:
  name: caphv-3r9grv-md-0-5nhvp
  namespace: caphv-54cb2q
Machine:
  name: caphv-3r9grv-md-0-595ccf6bf7x6656b-67rnr
  namespace: caphv-54cb2q
Cluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
HivelocityCluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
DeviceID: 14730
hostname: caphv-3r9grv-md-0-5nhvp.example.com
OsName: Ubuntu 20.x
script: |-
  #cloud-config
  ## template: jinja
  #cloud-config

  wr
ForceReload: true
---
level: INFO
time: "2023-11-20T07:57:21.866Z"
file: client/hvclient.go:178
message: ProvisionDevice() was successful (PutBareMetalDeviceIdResource)
controller: hivelocitymachine
controllerGroup: infrastructure.cluster.x-k8s.io
controllerKind: HivelocityMachine
HivelocityMachine:
  name: caphv-3r9grv-md-0-5nhvp
  namespace: caphv-54cb2q
namespace: caphv-54cb2q
name: caphv-3r9grv-md-0-5nhvp
reconcileID: 662fd9c2-e3f5-4af3-9796-6fd29b5999b7
HivelocityMachine:
  name: caphv-3r9grv-md-0-5nhvp
  namespace: caphv-54cb2q
Machine:
  name: caphv-3r9grv-md-0-595ccf6bf7x6656b-67rnr
  namespace: caphv-54cb2q
Cluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
HivelocityCluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
DeviceID: 14730
---
level: INFO
time: "2023-11-20T07:57:21.866Z"
file: device/state_machine.go:68
message: changing provisioning state
controller: hivelocitymachine
controllerGroup: infrastructure.cluster.x-k8s.io
controllerKind: HivelocityMachine
HivelocityMachine:
  name: caphv-3r9grv-md-0-5nhvp
  namespace: caphv-54cb2q
namespace: caphv-54cb2q
name: caphv-3r9grv-md-0-5nhvp
reconcileID: 662fd9c2-e3f5-4af3-9796-6fd29b5999b7
HivelocityMachine:
  name: caphv-3r9grv-md-0-5nhvp
  namespace: caphv-54cb2q
Machine:
  name: caphv-3r9grv-md-0-595ccf6bf7x6656b-67rnr
  namespace: caphv-54cb2q
Cluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
HivelocityCluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
caller: state-machine
old: provision-device
new: provisioned
---
level: INFO
time: "2023-11-20T08:16:44.278Z"
file: controllers/hivelocitycluster_controller.go:101
message: Starting reconciling cluster
controller: hivelocitycluster
controllerGroup: infrastructure.cluster.x-k8s.io
controllerKind: HivelocityCluster
HivelocityCluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
namespace: caphv-54cb2q
name: caphv-3r9grv
reconcileID: 62d47db5-642e-48b2-bb83-00ce74dcdfaf
HivelocityCluster:
  name: caphv-3r9grv
  namespace: caphv-54cb2q
```

**Comment by guettli** [comment]
@aniruddha2000 @janiskemper I guess it takes sometimes a bit longer. I created a new PR to increase the time the CI waits: https://github.com/hivelocity/cluster-api-provider-hivelocity/pull/122

**Comment by guettli** [comment]
@aniruddha2000 this PR was merged. Please rebase on the current main branch and try it again: https://github.com/hivelocity/cluster-api-provider-hivelocity/pull/122

**Comment by aniruddha2000** [comment]
@guettli I rebased but seems again failed

**Pull request merged**